src/fal/planner/schedule.py

```diff
@@ -0,0 +1,133 @@
+        assert kind is NodeKind.FAL_MODEL
+        task = FalModelTask(model_ids=model_ids)
+
+    model_node = node_graph.get_node(model_ids[-1])
+    assert model_node is not None
+
+    if isinstance(task, FalModelTask):
+        assert isinstance(model_node, DbtModelNode)
+        task.bound_model = model_node.model
+
+    post_hooks = [
+        FalHookTask(
+            hook_path=hook_path,
+            bound_model=model_node.model,
+        )
+        for hook_path in properties.get("post-hook", [])
+    ]
+    return Node(task, post_hooks=post_hooks)
+
+
+@dataclass
+class NodeQueue:
+    nodes: list[Node]
+    _staged_nodes: list[Node] = field(default_factory=list)
+    _counter: int = 0
+
+    def __bool__(self) -> bool:
+        return bool(self.nodes)
+
+    def _calculate_node_score(self, target_node: Node) -> tuple[int, int]:
+        # Determine the priority of the node by doing a bunch
+        # of calculations. This doesn't really need to be 100% precise,
+        # since if we don't have this we'll have to schedule randomly.
+
+        # 1. Number of nodes which are only waiting this node (direct dependants)
+        # 2. Number of nodes which are waiting this node and other nodes (indirect dependants)
+        # ...
+
+        direct_dependants = 0
+        indirect_dependants = 0
+
+        for node in self.nodes:
+            if node is target_node:
+                continue
+
+            if any(dependency is target_node for dependency in node.dependencies):
+                indirect_dependants += 1
+                if len(node.dependencies) == 1:
+                    direct_dependants += 1
+
+        return (direct_dependants, indirect_dependants)
+
+    def _stage_node(self, target_node: Node) -> None:
+        # When we start running a node, we don't want to simply remove it
+        # just yet (since that would unblock all of its dependencies). So
+        # we temporarily add it to a separate group of nodes (staged).
+
+        self._counter += 1
+        target_node.set_run_index(self._counter)
+
+        self.nodes.remove(target_node)
+        self._staged_nodes.append(target_node)
+
+    def finish(self, target_node: Node, status: int) -> None:
+        # When a staged node's execution is finished, we'll remove it
+        # alltogether and unblock all of its dependencies.
+        self._staged_nodes.remove(target_node)
+
+        if status == SUCCESS:
+            self._succeed(target_node)
+        else:
+            self._fail(target_node)
+
+        target_node.exit(status)
+
+    def _fail(self, target_node: Node) -> None:
+        for node in self.nodes.copy():
+            if target_node in node.dependencies:
+                self.nodes.remove(node)
+
+    def _succeed(self, target_node: Node) -> None:
+        for node in self.nodes.copy():
+            if target_node in node.dependencies:
+                node.dependencies.remove(target_node)
+
+    def iter_available_nodes(self) -> Iterator[Node]:
```

Comment:
Should this return a node instead of being an iterator of nodes? If it has a lot of unblocked nodes at the beginning without any depth, it will still process all of those.

```mermaid
graph LR;
  a;
  b --> c --> d --> e --> f --> g;
```

Would yield `b`, then `a`, and then need to be called again and get `c`, etc.

Reply:
I initially did it that way, but when I wrote the executor I've noticed that it can take as many as available so instead of calculating unblocked nodes at each call we only do calculate once and give up everything we can at the same time. If we did handle the number of jobs in the queue for the thread pool executor manually, I think your suggestion might be a decent optimization but currently it is all managed by the `concurrent.futures` so even if make this a method we'll still take all the unblocked nodes at each iteration and pass them at the same time.

Reply:
Maybe this could be a feature in the future though, with the improvements in the node scoring system (like timing info processing) it might matter a lot more. But for now, I'd abstain from that. WDYT?

Reply:
> I say change this not to a list of nodes but to return the most scored node and that is it.

But it really doesn't solve the problem with this system. If we change this to something that only returns a single node, then we'll port this logic back to the executor which will call this method as much as it can since we don't know how many available threads are there:

```py
while (node := scheduler.get_node()) is not None:
     submit_future_group(node)
```

which would be pretty much the same with the current approach, even a bit slower and a bit more complicated. Though maybe I am missing something.

Reply:
> then we'll port this logic back to the executor which will call this method as much as it can since we don't know how many available threads are there:
> ```
> while (node := scheduler.get_node()) is not None:
>      submit_future_group(node)
> ```
> which would be pretty much the same with the current approach, even a bit slower and a bit more complicated. Though maybe I am missing something.

Got it. I see your point! I think maybe a future change would be to make the executor aware of available threads, so it just calls when it makes sense. For now, let's leave it like it is!

Thanks for the explanation!